torch/export/export_schema.py

```diff
@@ -0,0 +1,319 @@
+    #   // external_data stores key-value pairs describing data location. Recognized keys are:
+    #   // - "location" (required) - POSIX filesystem path relative to the directory where the ONNX
+    #   //                           protobuf model was stored
+    #   // - "offset" (optional) - position of byte at which stored data begins. Integer stored as string.
+    #   //                         Offset values SHOULD be multiples 4096 (page size) to enable mmap support.
+    #   // - "length" (optional) - number of bytes containing data. Integer stored as string.
+    #   // - "checksum" (optional) - SHA1 digest of file specified in under 'location' key.
+    #   repeated StringStringEntryProto external_data
+    location: str
+    offset: str     # !!! Consider using int, but int has int_max limitation
+    length: str     # !!! Consider using int, but int has int_max limitation
+    checksum: str
+
+
+@dataclass
+class Storage:
+    class DataLocation(Enum):
+        Internal = auto()
+        External = auto()
+
+    data_location: DataLocation
+    data: Union[Buffer, ExternalBuffer]
+
+
+# This is a concrete tensor backed by storage
+@dataclass
+class Tensor:
+    # storage
+    storage: Storage
+
+    # metadata
+    meta: TensorMeta
+
+
+################################################################################
+# Following section is the defining the schema of 3 level construct: GraphModule, Graph, Node
+
+# IValue has no corresponding class in fx
+# IValue is the "values" that are passed between nodes in the graph
+# !!! Assumption: Only Tensor can be passed between nodes, and not other int/float/bool types...
+# IValue is a named virtual tensor, with an optional TensorMeta that describes the properties of the tensor
+# !!! Consider using a more descriptive name, e.g. TensorValue, TensorPlaceholder, TensorArgument, etc.
+@dataclass
+class IValue:
+    name: str   # A unique identifider name for the IValue
+                # The name will be used in the graph and node to refer to the IValue
+
+    meta: TensorMeta
+
+
+@dataclass
+class NodeMetadata:
+    stack_trace: str                      # source info of a node
+    nn_module_stack: str                  # stack of nn.Module that the node originates from
+    extra: Dict[str, str]                 # arbitrary string-string pairs for extra metadata
+
+
+# Maps to fx.Node
+@dataclass
+class Node:
+    op: str         # One of ['placeholder', 'call_function', 'get_attr', 'output'],
+                    # call_method and call_module are not supported, as they shouldn't apprear in the Caononical FX Graph
+                    # !!! Consider using an enum instead of string
+
+    target: str      # fully qualified name to the target, e.g. aten.add.Tensnor
+                     # !!! Consider using a structured operator name instead of string
+
+    args: List[Argument]              # args for this node
```

Comment:
@SherlockNoMad I think a unified "named argument" abstraction would suffice here.

Making everything a kwarg will simplify the operator call schema evolution very much, and now we only need to handle one `args` instead two `args` & `kwargs`.

I prefer to make named arguments ordered instead of a plain map:
```
args: List[Tuple[str, Argument]]
```

Reply:
@SherlockNoMad ^

Reply:
@suo wdyt

Reply:
I think it is probably nice to have a single named argument thing. The main thing I would worry about is that representing arguments differently from Python may cause us some headache later with operator signature evolution—but I can't think of a way it would bite us. And it's worth noting that even Python represents things with a [unified abstraction](https://docs.python.org/3.7/library/inspect.html?highlight=kwarg#inspect.Parameter).

Reply:
This is a discussion on function calling convention. Note that C++'s function arguments are purely positional. 

Imagine you are kernel writer, and you need to write a function signature for an op. Most likely, you will refer to the operator schema, and copy the arguments in the same order as specified in the operator schema. 

If the serialized arguments are stored in a different order, and identified by name, then downstream C++ consumer will need to convert the named argument to positional argument.

Reply:
I'd prefer to stick with Python conventions (so many details on ordering, defaults, *, what not...). 
C++ while positional has overloading, e.g.

Reply:
See current schema compat contract: https://github.com/pytorch/pytorch/blob/17bc40c19dd7ea24ff3e78ac52b92e93e555fca2/aten/src/ATen/core/function_schema.h#L283

Reply:
reproducing a comment on the internal discussion doc: 

Okay, I thought about this some more, and I think it does make sense to have a single abstraction for named arguments rather than splitting them into args and kwargs.

The reason here is somewhat subtle: what we're trying to conform should actually not be the Python calling convention. We should be trying to conform to the ATen specification, which today is represented by the [torchgen data model](https://github.com/pytorch/pytorch/blob/master/torchgen/model.py), which itself is centered around the JIT FunctionSchema representation.

This distinction is actually somewhat important, because the Python calling convention does not always conform to the ATen spec. For example today FX will populate kwargs with anything that was passed as a keyword argument, including things that are schematically positional arguments (since Python lets you specify named positional arguments in your function call).

So the safest thing to do is map directly onto FunctionSchema and/or torchgen constructs. If we map onto Python calling conventions, we will be forcing backends to do the same translation that TS/python_arg_parser do in order to map Python calls into FunctionSchema-complaint calls, which is complex and error-prone.

Reply:
yeah, I also think about our use cases and the way aten is fundamentally defined, and I think we should decouple from specific language calling conventions since we're talking about a neutral format which could be used in any language.